src/signatures.md

```diff
@@ -264,3 +264,73 @@ it should take `self` by value. Examples:
 
 * Closing a file and returning a result code.
 * A builder-pattern object which spits out the thing it was building. ([Example](https://docs.rs/bindgen/0.59.0/bindgen/struct.Builder.html#method.generate)).
+
+## Should I return an error, or panic?
+
+Panics should be used only for invariants, never for anything that you believe
+might happen. That's especially true for libraries.
+
+> Libraries which panic are super-rude and I hate them - MY
+
+Even in your own application code, panicking might not be wise:
+
+> Panicking in application logic for recoverable errors makes it way harder to librarify some code - AP
+
+If you really must have an API which can panic, add a `try_` equivalent too.
+
+## What should my error type be?
+
+[Rust's `Result` type](https://doc.rust-lang.org/std/result/) is parameterized
+over an error type. What should you use?
+
+For app code, consider [anyhow](https://docs.rs/anyhow/). For library code,
+use your own `enum` of error conditions - you can use [thiserror](https://docs.rs/thiserror/)
+to make this more pleasant.
+
+## When should I take or return `dyn Trait`?
+
+In either C++ or Rust, you can choose between monomorphization (that is, building
+code multiple times for each permutation of parameter types) or dynamic dispatch (i.e.
+looking up the correct implementation using vtables).
+
+In C++ the syntax is completely different - templates vs virtual functions.
+In Rust the syntax is almost identical - in some cases it's as simple as
+exchanging the `impl` keyword with the `dyn` keyword.
+
+Given this flexibility to switch strategies, which should you start with?
+
+In both languages, monomorphization tends to result in a quicker program (partly
+due to better inlining). It's arguably true that inlining is more important in
+Rust, due to its functional nature and pervasive use of iterators. Whether or
+not that's the reason, experienced Rustaceans usually start with `impl`:
+
+> It's best practice to start with monomorphization and move to `dyn`... - MG
+
+The main cost of monomorphization is larger binaries. There are cases where
+large amounts of code can end up being duplicated (the marvellous [serde](https://serde.rs/)
+is one).
+
+You _can_ choose to do things the other way round:
+
+> ... it’s workable practice to start with `dyn` and then move to `impl` when you have problems. - MG
+
+`dyn` can be awkward, and potentially expensive in different ways:
+
+> One thing to note about pervasive `dyn` is that because it unsizes the types it wraps, you need to box it if you want to store it by value. You end up with a good bit more allocator pressure if you try to have `dyn` field types. - AP
+
+## `<'a>`I seem to have lots of named lifetimes. Am `<'b>`I doing something wrong?
```

Comment:
Could possibly be merged with a similar FAQ entry in PR 16.

Reply:
Thanks, I'll consider that when I get to doing #16.